cmd/jujud/updateseries/updateseries.go

```diff
@@ -0,0 +1,396 @@
+	}
+	return lastError
+}
+
+func (c *UpdateSeriesCommand) createAgentConf(agentName string) (_ common.Conf, err error) {
+	defer func() {
+		if err != nil {
+			logger.Infof("Failed create agent conf for %s: %s", agentName, err)
+		}
+	}()
+
+	renderer, err := shell.NewRenderer("")
+	if err != nil {
+		return common.Conf{}, err
+	}
+
+	tag, err := names.ParseTag(agentName)
+	if err != nil {
+		return common.Conf{}, err
+	}
+	name := tag.Id()
+
+	var kind service.AgentKind
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		kind = service.AgentKindMachine
+	case names.UnitTagKind:
+		kind = service.AgentKindUnit
+	default:
+		return common.Conf{}, errors.NewNotValid(nil, fmt.Sprintf("agent %q is neither a machine nor a unit", agentName))
+	}
+
+	info := service.NewAgentInfo(
+		kind,
+		name,
+		c.dataDir,
+		paths.MustSucceed(paths.LogDir(c.toSeries)),
+	)
+	return service.AgentConf(info, renderer), nil
+}
+
+func (c *UpdateSeriesCommand) startAllAgents(ctx *cmd.Context) error {
+	running, err := sysdIsRunning()
+	switch {
+	case err != nil:
+		return err
+	case !running:
+		return errors.Errorf("systemd is not fully running, please reboot to start agents")
+	}
+
+	for _, unit := range c.unitAgents {
+		if err = c.startAgent(unit, service.AgentKindUnit); err != nil {
+			return errors.Annotatef(err, "failed to start %s service", "jujud-"+unit)
+		}
+		ctx.Infof("started %s service", "jujud-"+unit)
+	}
+
+	err = c.startAgent(c.machineAgent, service.AgentKindMachine)
+	if err == nil {
+		ctx.Infof("started %s service", "jujud-"+c.machineAgent)
+	}
+	return errors.Annotatef(err, "failed to start %s service", "jujud-"+c.machineAgent)
+}
+
+func (c *UpdateSeriesCommand) startAgent(name string, kind service.AgentKind) (err error) {
+	renderer, err := shell.NewRenderer("")
+	if err != nil {
+		return err
+	}
+	info := service.NewAgentInfo(
+		kind,
+		name,
+		c.dataDir,
+		paths.MustSucceed(paths.LogDir(c.toSeries)),
+	)
+	conf := service.AgentConf(info, renderer)
+	svcName := "jujud-" + name
+	svc, err := service.NewService(svcName, conf, c.toSeries)
+	if err = svc.Start(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *UpdateSeriesCommand) copyTools() (err error) {
+	defer func() {
+		if err != nil {
+			errors.Annotate(err, "failed to copy tools")
+		}
+	}()
+
+	// Get the current juju version from the machine agent
+	// conf file.
+	agentConf := agent.NewAgentConf(c.dataDir)
+	if err = agentConf.ReadConfig(c.machineAgent); err != nil {
+		return err
+	}
+	config := agentConf.CurrentConfig()
+	if config == nil {
+		return errors.Errorf("%s agent conf is not found", c.machineAgent)
+	}
+	jujuVersion := config.UpgradedToVersion()
+
+	fromVers := version.Binary{
+		Number: jujuVersion,
+		Arch:   arch.HostArch(),
+		Series: c.fromSeries,
+	}
+	toVers := version.Binary{
+		Number: jujuVersion,
+		Arch:   arch.HostArch(),
+		Series: c.toSeries,
+	}
+
+	// If tools with the new version don't already exist, copy old
```

Comment:
This makes for an interesting case where someone who want to add new units to an existing model will find they can't stay on the same version. Right? We're cheating in effect, with the assumption the agents will come up just fine on the newer version as-is. This will likely not always be true.

Reply:
Should we actually enforce series compatible tools as an init check to avoid this?

Reply:
We're only fudging the series name, not the juju version here.  I've been told that the only reason to have the series name in the tools directory name is historical.  That the same tools are used for every series.  This change is another reason to finally make the change official, no series name in the tools....  

New units in the same model will not be impacted, these changes are on an existing unit.  The user will also need tell juju that the machine is now running the new series.  And that new units will be created using the new series by default.   I'll have to check what will happen if you add a new unit to an updated machine... 

There is a current agent init check for juju tools in the correct version and unfortunately the expected series.  That was causing problems with the unit agent being re-written and started, part the abandoned PR did the same sort of copy to avoid the problem.

Re-reading the comment, it's a bit confusing, I'll update to reflect that only the series name is changing.